File: id/id/src/main/java/io/piano/android/id/PianoIdOAuthProvider.java
package io.piano.android.id;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;

public interface PianoIdOAuthProvider {

    /**
     * Get OAuth-provider name
     *
     * @return OAuth-provider name
     */
    @NonNull
    String getName();

    /**
     * Gets intent for starting sign in process
     *
     * @param context Context for building instance
     * @param extras  Bundle, which will be added to intent before start. You can modify, if you want
     * @return {@code Intent} instance for starting
     */
    @NonNull
    Intent buildIntent(@NonNull Context context, @NonNull Bundle extras);
}

File: id/id/src/main/java/io/piano/android/id/PianoIdOAuthProvider.kt
package io.piano.android.id

import android.content.Context
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultRegistryOwner
import io.piano.android.id.models.OAuthResult

public interface PianoIdOAuthProvider {
    /**
     * Get OAuth-provider name
     *
     * @return OAuth-provider name
     */
    public val name: String

    /**
     * Starts login process
     *
     * @param activity Activity, that started login process. You can use it as [Context] or [ActivityResultRegistryOwner]
     * @param clientId ClientId for OAuth provider, supplied by backend
     * @return successful result with token or cancelled result
     * @throws [Exception] if something goes wrong in login process
     */
    public suspend fun login(activity: ComponentActivity, clientId: String): OAuthResult
}
